### PR TITLE
Not allowed perform sensitive operations via gremlin

### PR DESCRIPTION
**Pull request by Linary**
Implement #145

Change-Id: I9a590fe40d3b5a808b569ed0af8fd83214a2941a

### CODE REVIEW

File: hugegraph-api/src/main/java/com/baidu/hugegraph/api/GremlinAPI.java [hunk 1]
```diff
@@ -120,6 +131,41 @@ public Response get(@Context HugeConfig conf,
...
         String query = uriInfo.getRequestUri().getRawQuery();
-        return doGetRequest(location, auth, query);
+        Response response = doGetRequest(location, auth, query);
+        return transformResponseIfNeed(response);
+    }
+
+    private static Response transformResponseIfNeed(Response response) {
+        int code = response.getStatusInfo().getStatusCode();
+        // No need to transform
+        if (code != Response.Status.INTERNAL_SERVER_ERROR.getStatusCode()) {
```
Comment (javeme) [inline hugegraph-api/src/main/java/com/baidu/hugegraph/api/GremlinAPI.java]: use response.getStatus()

File: hugegraph-api/src/main/java/com/baidu/hugegraph/api/GremlinAPI.java [hunk 2]
```diff
@@ -120,6 +131,41 @@ public Response get(@Context HugeConfig conf,
...
+        String exClassName = (String) map.get("Exception-Class");
+        @SuppressWarnings("unchecked")
+        List<String> exceptions = (List<String>) map.get("exceptions");
+        if (message == null || exClassName == null || exceptions == null) {
+            throw new IllegalStateException(String.format(
+                      "Invalid response for inner exception, should contains " +
+                      "Exception-Class, but got %s", map));
+        }
+        if (INNER_EXCEPTIONS.contains(exClassName)) {
+            String cause = !exceptions.isEmpty() ? exceptions.get(0) : "";
```
Comment (javeme) [inline hugegraph-api/src/main/java/com/baidu/hugegraph/api/GremlinAPI.java]: lost exceptions, maybe we can set to trace field

File: hugegraph-core/src/main/java/com/baidu/hugegraph/security/HugeSecurityManager.java [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+    @Override
+    public void checkPermission(Permission permission, Object context) {
+        // allow anything.
+    }
+
+    @Override
+    public void checkAccess(ThreadGroup g) {
+        if (this.callFromGremlin()) {
+            throw new SecurityException(
+                      "Not allowed to modify thread via gremlin");
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/security/HugeSecurityManager.java]: gremlin => Gremlin

File: hugegraph-core/src/main/java/com/baidu/hugegraph/security/HugeSecurityManager.java [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+//    }
+//
+//    @Override
+//    public void checkRead(String file, Object context) {
+//        if (this.callFromGremlin()) {
+//            throw new SecurityException("Not allowed to read file via gremlin");
+//        } else {
+//            super.checkRead(file, context);
+//        }
+//    }
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/security/HugeSecurityManager.java]: unused?

File: hugegraph-core/src/main/java/com/baidu/hugegraph/exception/SecurityException.java [hunk 5]
```diff
@@ -0,0 +1,20 @@
+/*
+ * Copyright (C) 2018 Baidu, Inc. All Rights Reserved.
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/exception/SecurityException.java]: update license

File: hugegraph-core/src/main/java/com/baidu/hugegraph/security/HugeSecurityManager.java [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+
+import org.slf4j.Logger;
+
+import com.baidu.hugegraph.util.Log;
+
+public class HugeSecurityManager extends SecurityManager {
+
+    private static final Logger LOG = Log.logger(HugeSecurityManager.class);
+
+    private static final String GremlinExecutor_Class =
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/security/HugeSecurityManager.java]: uppercase

File: hugegraph-dist/src/assembly/static/conf/hugegraph.policy [hunk 7]
```diff
@@ -0,0 +1,6 @@
+// Standard extensions get all permissions by default
+
+// 目录要使用变量
+grant codeBase "file:/Users/liningrui/IdeaProjects/baidu/xbu-data/hugegraph/hugegraph-0.8.0/hugegraph-core-0.8.0.jar" {
+    permission java.security.AllPermission;
+};
```
Comment (javeme) [inline hugegraph-dist/src/assembly/static/conf/hugegraph.policy]: add empty line

File: hugegraph-api/src/main/java/com/baidu/hugegraph/api/GremlinAPI.java [hunk 8]
```diff
@@ -120,6 +131,41 @@ public Response get(@Context HugeConfig conf,
...
+            String json = Json.createObjectBuilder()
+                              .add("exception", exClassName)
+                              .add("message", message)
+                              .add("cause", cause)
+                              .build().toString();
+            return Response.status(400)
```
Comment (javeme) [inline hugegraph-api/src/main/java/com/baidu/hugegraph/api/GremlinAPI.java]: Response.Status.XX_ERROR

**Comment by javeme** [comment]
only check in the gremlin threads, also consider asynchronous gremlin job

File: hugegraph-api/src/main/java/com/baidu/hugegraph/api/GremlinAPI.java [hunk 9]
```diff
@@ -120,6 +131,41 @@ public Response get(@Context HugeConfig conf,
...
+                              .add("exception", exClassName)
+                              .add("message", message)
+                              .add("cause", cause)
+                              .build().toString();
+            return Response.status(Response.Status.BAD_REQUEST)
+                           .type(MediaType.APPLICATION_JSON)
+                           .entity(json)
+                           .build();
+        } else {
+            return response;
```
Comment (javeme) [inline hugegraph-api/src/main/java/com/baidu/hugegraph/api/GremlinAPI.java]: return early

File: hugegraph-api/src/main/java/com/baidu/hugegraph/api/GremlinAPI.java [hunk 10]
```diff
@@ -120,6 +131,41 @@ public Response get(@Context HugeConfig conf,
...
+                      "Exception-Class, but got %s", map));
+        }
+        if (BAD_REQUEST_EXCEPTIONS.contains(exClassName)) {
+            String cause = !exceptions.isEmpty() ? exceptions.get(0) : "";
+            String json = Json.createObjectBuilder()
+                              .add("exception", exClassName)
+                              .add("message", message)
+                              .add("cause", cause)
+                              .build().toString();
+            return Response.status(Response.Status.BAD_REQUEST)
```
Comment (javeme) [inline hugegraph-api/src/main/java/com/baidu/hugegraph/api/GremlinAPI.java]: return 403 error

File: hugegraph-api/src/main/java/com/baidu/hugegraph/security/HugeSecurityManager.java [hunk 11]
```diff
@@ -0,0 +1,173 @@
...
+                      "Not allowed to read file via Gremlin");
+        } else {
+            super.checkRead(fd);
+        }
+    }
+
+    /**
+     * TODO: Open these will lead server can't start, because it need read
+     * conf files.
+     * Need to achieve no check at startup, open check after startup is complete
```
Comment (javeme) [inline hugegraph-api/src/main/java/com/baidu/hugegraph/security/HugeSecurityManager.java]: Can we distinguish if it was called from the script or from the code?

File: hugegraph-api/src/main/java/com/baidu/hugegraph/security/HugeSecurityManager.java [hunk 12]
```diff
@@ -161,7 +158,12 @@ public void checkConnect(String host, int port, Object context) {
     }
 
     private boolean callFromGremlin() {
-        StackTraceElement[] elements = Thread.currentThread().getStackTrace();
+        Thread curThread = Thread.currentThread();
+        if (!curThread.getName().startsWith(GREMLIN_SERVER_WORKER) &&
+            !curThread.getName().startsWith(TASK_WORKER)) {
+            return false;
```
Comment (javeme) [inline hugegraph-api/src/main/java/com/baidu/hugegraph/security/HugeSecurityManager.java]: move StackTrace here

File: hugegraph-dist/src/assembly/static/bin/hugegraph-server.sh [hunk 13]
```diff
@@ -79,6 +80,9 @@ fi
 # Execute the application and return its exit code
 ARGS="conf/gremlin-server.yaml conf/rest-server.properties"
 
-exec ${JAVA} -Dname="HugeGraphServer" -Dlog4j.configurationFile="${CONF}/log4j2.xml" \
+# Turn on security check
+exec ${JAVA} -Dname="HugeGraphServer" \
```
Comment (javeme) [inline hugegraph-dist/src/assembly/static/bin/hugegraph-server.sh]: switch by command line arg

File: hugegraph-dist/src/assembly/static/bin/start-hugegraph.sh [hunk 14]
```diff
@@ -1,19 +1,26 @@
...
 OPEN_MONITOR="false"
+OPEN_SECURITY_CHECK="false"
 VERBOSE=""
 SERVER_STARTUP_TIMEOUT_S=30
 
-while getopts "m:v" arg; do
+while getopts "m:s:v" arg; do
     case ${arg} in
         m) OPEN_MONITOR="$OPTARG" ;;
+        s) OPEN_SECURITY_CHECK="true" ;;
```
Comment (javeme) [inline hugegraph-dist/src/assembly/static/bin/start-hugegraph.sh]: OPEN_SECURITY_CHECK="$OPTARG"

File: hugegraph-dist/src/assembly/static/bin/start-hugegraph.sh [hunk 15]
```diff
@@ -1,19 +1,26 @@
 #!/bin/bash
 
 OPEN_MONITOR="false"
+OPEN_SECURITY_CHECK="false"
```
Comment (javeme) [inline hugegraph-dist/src/assembly/static/bin/start-hugegraph.sh]: enable security check by default and set value true

File: hugegraph-api/src/main/java/com/baidu/hugegraph/security/HugeSecurityManager.java [hunk 16]
```diff
@@ -0,0 +1,174 @@
...
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.baidu.hugegraph.security;
+
+import java.io.FileDescriptor;
+import java.security.Permission;
+
+public class HugeSecurityManager extends SecurityManager {
```
Comment (javeme) [inline hugegraph-api/src/main/java/com/baidu/hugegraph/security/HugeSecurityManager.java]: move to core

File: hugegraph-api/src/main/java/com/baidu/hugegraph/security/HugeSecurityManager.java [hunk 17]
```diff
@@ -0,0 +1,174 @@
...
+        } else {
+            super.checkExit(status);
+        }
+    }
+
+    @Override
+    public void checkRead(FileDescriptor fd) {
+        if (this.callFromGremlin()) {
+            throw new SecurityException(
+                      "Not allowed to read file via Gremlin");
```
Comment (zhoney) [inline hugegraph-api/src/main/java/com/baidu/hugegraph/security/HugeSecurityManager.java]: throw new SecurityException("Not allowed to read file via Gremlin");

File: hugegraph-api/src/main/java/com/baidu/hugegraph/api/gremlin/GremlinAPI.java [hunk 18]
```diff
@@ -121,6 +133,39 @@ public Response get(@Context HugeConfig conf,
...
+        String message = (String) map.get("message");
+        String exClassName = (String) map.get("Exception-Class");
+        @SuppressWarnings("unchecked")
+        List<String> exceptions = (List<String>) map.get("exceptions");
+        if (message == null || exClassName == null || exceptions == null) {
+            throw new IllegalStateException(String.format(
+                      "Invalid response for inner exception, should contains " +
+                      "Exception-Class, but got %s", map));
+        }
+        if (!BAD_REQUEST_EXCEPTIONS.contains(exClassName)) {
```
Comment (javeme) [inline hugegraph-api/src/main/java/com/baidu/hugegraph/api/gremlin/GremlinAPI.java]: FORBIDDEN_REQUEST_EXCEPTIONS

File: hugegraph-api/src/main/java/com/baidu/hugegraph/api/gremlin/GremlinAPI.java [hunk 19]
```diff
@@ -121,6 +133,39 @@ public Response get(@Context HugeConfig conf,
...
+            return response;
+        }
+        String json = Json.createObjectBuilder()
+                          .add("exception", exClassName)
+                          .add("message", message)
+                          .add("trace", JsonUtil.toJson(exceptions))
+                          .build().toString();
+        return Response.status(Response.Status.FORBIDDEN)
```
Comment (javeme) [inline hugegraph-api/src/main/java/com/baidu/hugegraph/api/gremlin/GremlinAPI.java]: consistent format for response

File: hugegraph-core/src/main/java/com/baidu/hugegraph/security/HugeSecurityManager.java [hunk 20]
```diff
@@ -69,10 +70,20 @@ public void checkExit(int status) {
         }
     }
 
+    @Override
+    public void checkExec(String cmd) {
+        if (this.callFromGremlin()) {
+            throw new SecurityException(
+                      "Not allowed to exec command via Gremlin");
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/security/HugeSecurityManager.java]: execute

File: hugegraph-core/src/main/java/com/baidu/hugegraph/security/HugeSecurityManager.java [hunk 21]
```diff
@@ -69,10 +70,20 @@ public void checkExit(int status) {
...
+        } else {
+            super.checkExec(cmd);
+        }
+    }
+
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/security/HugeSecurityManager.java]: add methods:
```
checkCreateClassLoader() // load class
checkLink(String lib) // load jar

checkPrintJobAccess()
checkSystemClipboardAccess()

checkPropertiesAccess() // env property
checkPropertyAccess(String key)

checkSetFactory() // socket
```
https://docs.oracle.com/javase/7/docs/api/java/lang/SecurityManager.html

File: hugegraph-core/src/main/java/com/baidu/hugegraph/security/HugeSecurityManager.java [hunk 22]
```diff
@@ -156,6 +176,15 @@ public void checkConnect(String host, int port, Object context) {
...
 
+    @Override
+    public void checkMulticast(InetAddress maddr) {
+        if (this.callFromGremlin()) {
+            throw new SecurityException("Not allowed to multicast via Gremlin");
+        } else {
+            super.checkMulticast(maddr);
+        }
+    }
+
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/security/HugeSecurityManager.java]: add `checkMulticast(InetAddress maddr, byte ttl)`

File: hugegraph-test/src/main/java/com/baidu/hugegraph/unit/core/SecurityManagerTest.java [hunk 23]
```diff
@@ -0,0 +1,178 @@
...
+    public void testProperties() {
+        String result = runGremlinJob("System.getProperties()");
+        assertError(result,
+                    "Not allowed to access system properties via Gremlin");
+
+        result = runGremlinJob("System.getProperty(\"java.version\")");
+        assertError(result,
+                    "Not allowed to access system property(java.version) " +
+                    "via Gremlin");
+    }
```
Comment (javeme) [inline hugegraph-test/src/main/java/com/baidu/hugegraph/unit/core/SecurityManagerTest.java]: add more test

File: hugegraph-test/src/main/java/com/baidu/hugegraph/unit/core/SecurityManagerTest.java [hunk 24]
```diff
@@ -83,73 +101,185 @@ public void testExit() {
 
     @Test
     public void testFile() {
-        String result = runGremlinJob("fr = new FileReader(new File(\"\"));" +
-                                      "fr.read();fr.close()");
+        // read file
+        try (FileInputStream fis = new FileInputStream(new File(""))) {
+        } catch (IOException ignored) {
+        }
```
Comment (javeme) [inline hugegraph-test/src/main/java/com/baidu/hugegraph/unit/core/SecurityManagerTest.java]: move to the previous line

File: hugegraph-test/src/main/java/com/baidu/hugegraph/unit/core/SecurityManagerTest.java [hunk 25]
```diff
@@ -0,0 +1,320 @@
...
+
+        sm.checkMulticast(InetAddress.getByAddress(new byte[]{0, 0, 0, 0}));
+        result = runGremlinJob("bs = [0, 0, 0, 0] as byte[];" +
+                               "System.getSecurityManager()" +
+                               ".checkMulticast(InetAddress.getByAddress(bs))");
+        assertError(result, "Not allowed to multicast via Gremlin");
+
+        sm.checkMulticast(InetAddress.getByAddress(new byte[]{0, 0, 0, 0}),
+                                                   (byte) 1);
+        result = runGremlinJob("bs = [0, 0, 0, 0] as byte[];ttl = (byte) 1;" +
```
Comment (javeme) [inline hugegraph-test/src/main/java/com/baidu/hugegraph/unit/core/SecurityManagerTest.java]: add a space before ttl

File: hugegraph-test/src/main/java/com/baidu/hugegraph/unit/core/SecurityManagerTest.java [hunk 26]
```diff
@@ -0,0 +1,320 @@
...
+
+        sm.checkRead("", new Object());
+        result = runGremlinJob("System.getSecurityManager()" +
+                               ".checkRead(\"\", new Object())");
+        assertError(result, "Not allowed to read file via Gremlin");
+
+        // write file
+        try (FileOutputStream fos = new FileOutputStream(new File(""))) {
+        } catch (IOException ignored) {
+        }
```
Comment (javeme) [inline hugegraph-test/src/main/java/com/baidu/hugegraph/unit/core/SecurityManagerTest.java]: ditto

File: hugegraph-test/src/main/java/com/baidu/hugegraph/unit/core/SecurityManagerTest.java [hunk 27]
```diff
@@ -0,0 +1,311 @@
...
+        assertError(result, "Not allowed to multicast via Gremlin");
+
+        sm.checkSetFactory();
+        result = runGremlinJob("System.getSecurityManager().checkSetFactory()");
+        assertError(result, "Not allowed to set socket factory via Gremlin");
+    }
+
+    @Test
+    public void testExec() throws IOException {
+        Runtime.getRuntime().exec("ls");
```
Comment (zhoney) [inline hugegraph-test/src/main/java/com/baidu/hugegraph/unit/core/SecurityManagerTest.java]: same command is better

File: hugegraph-test/src/main/java/com/baidu/hugegraph/unit/core/SecurityManagerTest.java [hunk 28]
```diff
@@ -0,0 +1,311 @@
...
+        assertError(result, "Not allowed to print job via Gremlin");
+    }
+
+    @Test
+    public void testSystemClipboardAccess() {
+        sm.checkSystemClipboardAccess();
+        String result = runGremlinJob("System.getSecurityManager()" +
+                                      ".checkSystemClipboardAccess()");
+        assertError(result,
+                    "Not allowed to access sysytem clipboard via Gremlin");
```
Comment (zhoney) [inline hugegraph-test/src/main/java/com/baidu/hugegraph/unit/core/SecurityManagerTest.java]: sysytem -> system

File: hugegraph-core/src/main/java/com/baidu/hugegraph/security/HugeSecurityManager.java [hunk 29]
```diff
@@ -0,0 +1,348 @@
...
+    public void checkSystemClipboardAccess() {
+        if (callFromGremlin()) {
+            throw new SecurityException(
+                      "Not allowed to access sysytem clipboard via Gremlin");
+        }
+        super.checkSystemClipboardAccess();
+    }
+
+    @Override
+    public void checkPackageAccess(String pkg) {
```
[the remark this answers is not in the document]
Comment (Linary) [reply hugegraph-core/src/main/java/com/baidu/hugegraph/security/HugeSecurityManager.java]: just override all methods explicitly, avoid dangers caused by omissions

**Pull request merged**